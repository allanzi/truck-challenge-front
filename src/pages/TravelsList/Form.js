/* eslint-disable no-undef */

import React, { useState, useEffect } from 'react';
import {
  Modal, Form, Input, Spin, Row, Col, Alert, InputNumber, message
} from 'antd';

import truckApi from '../../services/api.service';

const TravelFormPage = ({ params, isUpdate }) => {
  const [loading, setLoading] = useState(false);
  const [showErrorAlert, setShowErrorAlert] = useState(false);
  const [form] = Form.useForm();

  // eslint-disable-next-line consistent-return
  const getTravel = async () => {
    try {
      setLoading(true);
      const { data: { data } } = await truckApi.showTravel(params.id);
      form.setFieldsValue({
        user_id: data.user_id,
        from_latitude: data.from.latitude,
        from_longitude: data.from.longitude,
        to_latitude: data.to.latitude,
        to_longitude: data.to.longitude
      });
    } catch (e) {
      if (isUpdate) {
        return message.error('Não foi possível editar essa viagem! Tente novamente mais tarde.');
      }

      message.error('Não foi possível cadastrar essa viagem! Tente novamente mais tarde.');
    } finally {
      setLoading(false);
    }
  };

  const handleRequest = async () => {
    try {
      const values = await form.validateFields();
      const data = {
        user_id: values.user_id,
        from_: {
          latitude: values.from_latitude,
          longitude: values.from_longitude,
        },
        to: {
          latitude: values.to_latitude,
          longitude: values.to_longitude,
        }
      };

      if (isUpdate) {
        await truckApi.updateTravel(params.id, data);
        message.success('Viagem atualizada com sucesso');
      } else {
        await truckApi.createTravel(data);
        message.success('Viagem criada com sucesso');
      }

      // eslint-disable-next-line no-return-assign
      setTimeout(() => window.location.pathname = '/travels', 1500);
    } catch (e) {
      message.error('Corrija os campos!');

      if (e.message === 'Request failed with status code 400') {
        setShowErrorAlert(true);
      }
    }
  };

  const handleCancel = () => {
    window.location.pathname = '/travels';
  };

  useEffect(() => {
    if (isUpdate) {
      getTravel();
    }

    document.title = 'Truck Panel | Viagens';
  }, []);

  return (
    <Modal
      visible
      title={isUpdate ? 'Editar Viagem' : 'Cadastrar Viagem'}
      okText={isUpdate ? 'Atualizar' : 'Cadastrar'}
      cancelText="Cancelar"
      onOk={handleRequest}
      onCancel={handleCancel}
    >
      {loading ? (
        <Row justify="center">
          <Col>
            <br />
            <br />
            <br />
            <Spin />
            <br />
            <br />
            <br />
          </Col>
        </Row>
      ) : (
        <>
          {showErrorAlert && (
          <>
            <Alert
              message="Campo Caminhoneiro ID inválido!"
              description="Deve ser um ObjectID válido ou caminhoneiro não encontrado!"
              type="error"
              closable
            />
            <br />
          </>
          )}
          <Form
            form={form}
            layout="vertical"
            name="edit_travel"
          >
            <Form.Item
              rules={[{ required: true, message: 'Este campo é obrigatório!' }]}
              name="user_id"
              label="Caminhoneiro ID"
            >
              <Input />
            </Form.Item>
            <Row gutter={[8, 8]}>
              <Col span={12}>
                <Form.Item
                  rules={[{ required: true, message: 'Este campo é obrigatório!' }]}
                  name="from_latitude"
                  label="Origem latitude"
                >
                  <InputNumber
                    style={{
                      width: '100%',
                    }}
                  />
                </Form.Item>
              </Col>
              <Col span={12}>
                <Form.Item
                  rules={[{ required: true, message: 'Este campo é obrigatório!' }]}
                  name="from_longitude"
                  label="Origem longitude"
                >
                  <InputNumber
                    style={{
                      width: '100%',
                    }}
                  />
                </Form.Item>
              </Col>
            </Row>
            <Row gutter={[8, 8]}>
              <Col span={12}>
                <Form.Item
                  rules={[{ required: true, message: 'Este campo é obrigatório!' }]}
                  name="to_latitude"
                  label="Destino latitude"
                >
                  <InputNumber
                    style={{
                      width: '100%',
                    }}
                  />
                </Form.Item>
              </Col>
              <Col span={12}>
                <Form.Item
                  rules={[{ required: true, message: 'Este campo é obrigatório!' }]}
                  name="to_longitude"
                  label="Destino longitude"
                >
                  <InputNumber
                    style={{
                      width: '100%',
                    }}
                  />
                </Form.Item>
              </Col>
            </Row>
          </Form>
        </>
      )}
    </Modal>
  );
};

export default TravelFormPage;
