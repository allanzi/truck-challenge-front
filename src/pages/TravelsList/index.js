import React, { useEffect, useState } from 'react';
import {
  Breadcrumb, Table, Tag, Popconfirm, message, Button, Row, Col
} from 'antd';

import truckApi from '../../services/api.service';
import './styles.css';
import TravelForm from './Form';

const TravelsListPage = ({ match: { params } }) => {
  const [travels, setTravels] = useState({ loading: true, data: [] });
  const [visibleForm, setVisibleForm] = useState(false);
  const [isUpdate, setIsUpdate] = useState(true);

  const getTravels = async () => {
    const { data: { data } } = await truckApi.listTravels();
    const transformedData = [];

    data.map(async (travel) => {
      // eslint-disable-next-line no-underscore-dangle
      const { data: { data: { name } } } = await truckApi.showUser(travel.user_id);

      transformedData.push({ name, ...travel });
    });

    setTimeout(() => setTravels({ loading: false, data: transformedData }), 500);
  };

  const handleDelete = async (id) => {
    try {
      await truckApi.deleteTravel(id);
      message.success('Excluído com sucesso');
      getTravels();
    } catch {
      message.error('Não foi possível excluir, tente novamente mais tarde.');
    }
  };

  const handleEditButton = (id) => {
    window.location.pathname = `/travels/${id}`;
  };

  const handleCreateButton = () => {
    window.location.pathname = '/travels/create';
  };

  useEffect(() => {
    getTravels();
    document.title = 'Truck Panel | Viagens';

    if ('id' in params && params.id) {
      if (params.id === 'create') {
        setIsUpdate(false);
      }

      setVisibleForm(true);
    }
  }, []);

  return (
    <>
      <Row justify="space-between" align="middle">
        <Col span={12}>
          <Breadcrumb style={{ margin: '16px 0' }}>
            <Breadcrumb.Item>Truck Panel</Breadcrumb.Item>
            <Breadcrumb.Item>Viagens</Breadcrumb.Item>
          </Breadcrumb>
        </Col>
        <Col spam={12}>
          <Button type="primary" onClick={handleCreateButton}>Cadastrar viagem</Button>
        </Col>
      </Row>
      <div className="site-layout-background" style={{ padding: 24, minHeight: 360 }}>
        <Table
          rowKey={(item) => item.name}
          loading={travels.loading}
          dataSource={travels.data}
          locale={{
            emptyText: 'Sem dados para exibir',
          }}
          columns={[
            {
              title: 'Nome',
              dataIndex: 'name',
              key: 'name',
            },
            {
              title: 'Origem',
              dataIndex: 'from',
              key: 'from',
              render: (from) => (
                <>
                  <span>
                    <Tag color="green">
                      {from.latitude}
                    </Tag>
                  </span>
                  <span>
                    <Tag color="green">
                      {from.longitude}
                    </Tag>
                  </span>
                </>
              )
            },
            {
              title: 'Destino',
              dataIndex: 'to',
              key: 'to',
              render: (to) => (
                <>
                  <span>
                    <Tag color="green">
                      {to.latitude}
                    </Tag>
                  </span>
                  <span>
                    <Tag color="green">
                      {to.longitude}
                    </Tag>
                  </span>
                </>
              )
            },
            {
              title: 'Ações',
              key: 'action',
              render: (text, record) => (
                <span>
                  <a style={{ marginRight: 16 }} onClick={() => handleEditButton(record.id)}>Editar</a>
                  <Popconfirm
                    title="Realmente deseja excluir?"
                    onConfirm={() => handleDelete(record.id)}
                    okText="Sim"
                    cancelText="Não"
                  >
                    <a href="javascript:void(0);">Excluir</a>
                  </Popconfirm>
                </span>
              ),
            },
          ]}
        />
        {visibleForm && (
          <TravelForm
            params={params}
            isUpdate={isUpdate}
          />
        )}
      </div>
    </>
  );
};

export default TravelsListPage;
