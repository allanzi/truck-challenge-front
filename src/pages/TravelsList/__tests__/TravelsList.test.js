import React from 'react';

import TravelsList from '../index';

it('should render correct', () => {
  const wrapper = shallow(<TravelsList match={{ params: { id: 'create' } }} />);

  expect(wrapper).toMatchSnapshot();
});
