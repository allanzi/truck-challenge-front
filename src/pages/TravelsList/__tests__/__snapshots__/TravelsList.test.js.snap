// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correct 1`] = `
<Fragment>
  <Row
    align="middle"
    gutter={0}
    justify="space-between"
  >
    <Col
      span={12}
    >
      <Breadcrumb
        separator="/"
        style={
          Object {
            "margin": "16px 0",
          }
        }
      >
        <BreadcrumbItem
          separator="/"
        >
          Truck Panel
        </BreadcrumbItem>
        <BreadcrumbItem
          separator="/"
        >
          Viagens
        </BreadcrumbItem>
      </Breadcrumb>
    </Col>
    <Col
      spam={12}
    >
      <Button
        block={false}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="primary"
      >
        Cadastrar viagem
      </Button>
    </Col>
  </Row>
  <div
    className="site-layout-background"
    style={
      Object {
        "minHeight": 360,
        "padding": 24,
      }
    }
  >
    <Table
      columns={
        Array [
          Object {
            "dataIndex": "name",
            "key": "name",
            "title": "Nome",
          },
          Object {
            "dataIndex": "from",
            "key": "from",
            "render": [Function],
            "title": "Origem",
          },
          Object {
            "dataIndex": "to",
            "key": "to",
            "render": [Function],
            "title": "Destino",
          },
          Object {
            "key": "action",
            "render": [Function],
            "title": "Ações",
          },
        ]
      }
      dataSource={Array []}
      loading={true}
      locale={
        Object {
          "emptyText": "Sem dados para exibir",
        }
      }
      rowKey={[Function]}
    />
  </div>
</Fragment>
`;
